Inventory: only pick up stacks from the same list

SmallJoker opened the pull request:
Fixes #13961

This is an alternative proposal for #14656

## To do

This PR is Ready for Review.

## How to test

See #13961

grorp reviewed (changes requested):
- This PR fixes the problematic behavior in MTG's creative inventory, but ...

 - I can trivially restore the bug by making the "normal inventory" used for testing part of the same inventory list as the "creative inventory": https://github.com/grorp/minetest_game/commit/bdf46b83932ba6b07259bd7b3ae9352132199bf7
    
    https://github.com/minetest/minetest/assets/82708541/2ef28891-857e-420e-969c-d7395a7d6021

    so this doesn't seem to fix the underlying issue

- This PR also prohibits valid usage of "drag to pick up", e.g. across the "main" and the "craft" player inventory. Video taken on master branch, no longer works with PR:
    
    https://github.com/minetest/minetest/assets/82708541/b30d2658-e625-456b-a1fe-56f4355e87c2

SmallJoker commented:
Problem analysis:

1. The code block in `} else if (pickup_amount > 0) {` tries to move the hovered (not selected!) ItemStack to the selected ItemStack index to combine and re-select the sum again (="pickup").
2. This fails because the creative inventory always sets the stack to count=1
3. The previously selected index is still selected, thus this bug.

The solution is to instead move the pickup ItemStack *TO* the hovered ItemStack.

Problem: we expect the exact opposite when holding a stack (not dragging) and picking up additional items using the mouse wheel. I do have a code approach that would fix it, but I am afraid that it would be a similar jumpscare as the other PR.

grorp reviewed (changes requested):
Drag & drop is broken when the destination slot is not empty (because of drag & drop not being allowed with source slot == destination slot)

https://github.com/minetest/minetest/assets/82708541/77584fba-7a13-480c-896d-afeb0e550d13

Besides, there are still glitches:

https://github.com/minetest/minetest/assets/82708541/0fcfd17b-2194-40a1-8f3e-c7e2ba2b81b8

and when `allow_take` returns 0:

https://github.com/minetest/minetest/assets/82708541/ceb945d3-f4d7-4a62-be78-53db94ff5ae9

grorp commented:
It seems to me that which slot is being moved and which stays isn't actually the problem here. The main problem seems to be that the client-side selection ignores when the server disallows a pick-up.

SmallJoker commented:
> Drag & drop is broken when the destination slot is not empty [..]

I did have a fix for that (before you tested the PR), but the changes became even messier. Hence I converted this PR to a draft to re-think. The `pickup_amount` parameter is in fact not suited to implement the stack movement that I thought of.

> The main problem seems to be that the client-side selection ignores when the server disallows a pick-up.

That is an explicit case that is ignored by `updateSelectedItem`, which was initially modified by #14656. It is very difficult to take both into consideration: inventory lag (latency) and disallowed (or restricted) item movements. As of now I have no good idea on how to approach this particular issue.

grorp commented:
The original bug case now works, although a bit weird:

https://github.com/minetest/minetest/assets/82708541/ea9c92e7-7966-429e-97e3-1276cd3d385c

However, the glitches shown above are still there. This new bug introduced by this PR is basically the opposite of the original bug:

https://github.com/minetest/minetest/assets/82708541/fb0ab61d-846e-468a-ad10-dd9942b4d09d

The behavior in that case before this PR was more reasonable: Dragging over the creative slot would simply increase the stack count by one.

SmallJoker commented:
> Drag & drop is broken when the destination slot is not empty 

Fixed.

> Besides, there are still glitches:

This performs a move request because you're collecting one stack from the creative inventory over and over again, which is technically what the drag & pickup feature is supposed to do.

> The original bug case now works, although a bit weird:

It works like any other item stack combining actions, with the difference that the creative inventory is re-sent by the server after `on_take`.

> However, the glitches shown above are still there. This new bug introduced by this PR is basically the opposite of the original bug:

Fixed by assuming that moving the entire stack is possible, hence deselecting it. This prediction however might result in more annoying handling when `on_put` actions are restricted.

grorp commented:
> This performs a move request because you're collecting one stack from the creative inventory over and over again, which is technically what the drag & pickup feature is supposed to do.

What I meant by glitch is this state (screenshot of the relevant video):

![screenshot of video](https://github.com/minetest/minetest/assets/82708541/27cbbf6f-c473-4d36-9e59-7816ce64a83c)

Two stacks are shown but one of them is actually just one item. This still happens with the latest version of this PR.

> Fixed by assuming that moving the entire stack is possible, hence deselecting it.

Releasing the stack in the case shown in the relevant video now actually puts it down, is that what you mean?

SmallJoker commented:
> What I meant by glitch is this state (screenshot of the relevant video):

This is what happens internally:

1. stack of 10 is picked up (selected item)
2. +=1 from the creative inventory (selected item, prediction)
3. move action is sent
4. server corrects the slot to count = 1 (client still recognizes it as selected)
5. the prediction does not know whether all actions are complete, thus keeping the selected stack unmodified

From how I understood the code, this would require some sort of "tracking number" for the total count of inventory actions from a player (client -> server -> client confirmation), so that the stack can be forced to update when the client knows that all actions were handled (i.e. we're in sync).
Out of scope for this PR.

> Releasing the stack in the case shown in the relevant video now actually puts it down, is that what you mean?

Exactly. This is in fact very closely related to the point above. The server reverts both inventories due to disallowed `on_put`, leaving the client-side prediction in a state where it thinks there's more stacks coming that it could put into the destination stack. 

A possible solution might be to set `m_selected_amount = move_amount` after L4765. That would however disable the prediction, which is not very good either in case of laggy servers (where the inventory has yet not been re-sent).

nininik0 commented:
This should be enable-able,  as I find this useful on Android, we can't just hold shift and tap items to get many from a chest of everything, and i usually do this to get a certain item amount without lag delay when i have to tap the item from the chest, then tap my inventory, doing this a lot in fast intervals causes a insane lag delay. However when using this weird inventory behavior, it does not lag and is actually useful to me and other Android players.

nininik0 commented:
As an android player, I find this useful as when on Android using an inventory like chest of everything and not creative, we want to get a certain amount we use the inventory glitch, instead of having to tap the item and then tap the inventory. It uses less time and is less laggy, as tapping the inventory and the item many times fast causes a huge lag delay. That won't go away until all the items go into the inv how many times you tapped. I suggest making this be Optionally enable-able.

nininik0 commented:
Oops sorry commented twice (delete this one and the one above)

grorp reviewed (commented):
Tested again. This PR fixes one bug, but introduces a different bug, which is basically the exact opposite of the old bug.

While "dragging items over the MTG creative inventory" working correctly is not as important as "dragging&dropping items from the MTG creative inventory", this makes this fix look like it's not addressing the root cause, but instead trading one bad case for another. (Also, there could possibly be other situations in other inventories as well where the new bug is exhibited.)

master branch, dragging from MTG creative inventory (buggy)

https://github.com/user-attachments/assets/769618fe-6344-4213-8991-4a740b267408

this PR, dragging from MTG creative inventory (working at least similarly to what I would have expected)

https://github.com/user-attachments/assets/aee08cca-048a-4005-bb83-76e314d7de02

master branch, dragging over MTG creative inventory (working as expected (consider the "pickup when dragging over item" feature))

https://github.com/user-attachments/assets/8cb8539b-2cca-433e-988e-69e8e319220a

this PR, dragging over MTG creative inventory (buggy)

https://github.com/user-attachments/assets/1bb9e6fc-2213-4c9c-bfe8-a04c2a6c4f6a

SmallJoker commented:
Thank you for the review. It's been a bit too long since I last worked with this code. It's a code region filled with pitfalls and thus, I am not surprised that there's another bug rising from the ashes.
As of now I cannot think of any sane solution to fix this stack prediction issue. Closing until I'm motivated again.
